### PR TITLE
refactor(tiktok): Update TikTok provider configuration with custom Fetch

### PR DESCRIPTION
**Pull request by DewinU**
## ☕️ Reasoning

- Refactor Tiktok Provider default configuration to use the new customFetch feature #11975 
- Added more documentation on how to setup this provider using de default configuration and also as a custom provider on your own.

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

- #11468 

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

# Video Demostration
https://youtu.be/M-PCU-e8IGI?si=uunGbOxhNXIRlnXT

# Custom Provider
You dont really need to wait for this PR to be merged, if you already need the Tiktok Provider on your application, you can set it up as a custom provider

```ts
const CustomTiktok: OAuth2Config<any> & Provider =
{
  async [customFetch](...args) {
    const url = new URL(args[0] instanceof Request ? args[0].url : args[0])
    if (url.pathname.endsWith("/token/")) {
      const [url, request] = args

      const customHeaders = {
        ...request?.headers,
        "content-type": "application/x-www-form-urlencoded",
      }

      const customBody = new URLSearchParams(request?.body as string)
      customBody.append("client_key", process.env.AUTH_TIKTOK_ID!)
      const response = await fetch(url, {
        ...request,
        headers: customHeaders,
        body: customBody.toString(),
      })
      const json = await response.json()
      return Response.json({ ...json })
    }
    return fetch(...args)
  },
  id: "tiktok",
  name: "TikTok",
  type: "oauth",
  client: {
    token_endpoint_auth_method: "client_secret_post",
  },
  authorization: {
    url: "https://www.tiktok.com/v2/auth/authorize",
    params: {
      client_key: process.env.AUTH_TIKTOK_ID,
      scope: "user.info.profile",
    },
  },

  token: "https://open.tiktokapis.com/v2/oauth/token/",
  userinfo:
    "https://open.tiktokapis.com/v2/user/info/?fields=open_id,avatar_url,display_name,username",

  profile(profile) {
    return {
      id: profile.data.user.open_id,
      name: profile.data.user.display_name,
      image: profile.data.user.avatar_url,
      email: profile.data.user.email || profile.data.user.username || null,
    }
  },
}
```

and then add it at your list of providers

```ts
export const { handlers, auth } = NextAuth({ providers: [ GitHub, CustomTiktok ] })
```
Any feedback will be greatly appreciated.

### CODE REVIEW

**Comment by mattstrick** [comment]
Looking forward to this update!

File: packages/core/src/providers/tiktok.ts [hunk 1]
```diff
@@ -153,13 +155,15 @@ export interface TiktokProfile extends Record<string, any> {
  *  - [TikTok app console](https://developers.tiktok.com/)
  *  - [TikTok login kit documentation](https://developers.tiktok.com/doc/login-kit-web/)
  *  - [Avaliable Scopes](https://developers.tiktok.com/doc/tiktok-api-scopes/)
+ *  - [Sandbox for testing](https://developers.tiktok.com/blog/introducing-sandbox)
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.ts:158]: This is huge! 🤯 Finally!

**Review by balazsorban44 — APPROVED** [review]

**Comment by dragidavid** [comment]
Hey folks, was about to create a new issue when I found this. Using the latest version (`5.0.0-beta.25`) the TikTok provider is still throwing the `The request parameters are malformed.` error on sign in. 

Funny enough with the code snippet above, it all works. Not sure if that was included in the v5 beta?

**Comment by DewinU** [comment]
Hey @dragidavid, you're right—the PR was merged into the main branch, but it hasn't been shipped yet since there hasn't been a new release of Auth.js since November 2024. I believe the maintainers are being extra cautious about compatibility and testing all providers with the new APIs.

In the meantime, you can implement the TikTok provider as a Custom Provider for your specific use case. Let me know if you have any questions about its setup or usage!

**Comment by dragidavid** [comment]
> Hey @dragidavid, you're right—the PR was merged into the main branch, but it hasn't been shipped yet since there hasn't been a new release of Auth.js since November 2024. I believe the maintainers are being extra cautious about compatibility and testing all providers with the new APIs.
> 
> In the meantime, you can implement the TikTok provider as a Custom Provider for your specific use case. Let me know if you have any questions about its setup or usage!

Yea there is an issue where you want to grab more things in the `profile`. Even after changing the `userInfo` endpoint with the fields that you want and seeing them actually being returned (using `debug). It still won't put them into the session 🤔

**Comment by DewinU** [comment]
If you want to add them to session you need to add them manually like a Role (https://authjs.dev/guides/role-based-access-control?framework=next-js)

Heres some code reference
```ts
const CustomTiktok: OAuth2Config<any> & Provider =
{
  /* Configuration Code  */
  userinfo:
    "https://open.tiktokapis.com/v2/user/info/?fields=open_id,avatar_url,display_name,username,profile_web_link,profile_deep_link,is_verified",

  profile(profile) {
    
    return {
      id: profile.data.user.open_id,
      name: profile.data.user.display_name,
      image: profile.data.user.avatar_url,
      email: profile.data.user.email || profile.data.user.username || null,
      ...profile.data.user, // --> any other fields provided by the provider
    }
  },
}
```
You add in profile any other fields provided by the profile object, then you use the callbacks like `signIn`, `jwt` and `session`

```ts
export const { handlers, auth, signIn, signOut } = NextAuth({
  providers: [
    CustomTiktok
  ],
  basePath: "/auth",
  session: { strategy: "jwt" },
  callbacks: {
    async signIn({ account, user, profile, email, credentials}){
      //Do Whatever you want with the user object

      return true
    },
    jwt({ token, trigger, session, account, user, profile }) {
      if(user){
        token.is_verified = user.is_verified,
        token.profile_deep_link = user.profile_deep_link,
        token.profile_web_link = user.profile_web_link
      }
      return token
    },
    async session({ session, token, user }) {
      if (token?.accessToken) session.accessToken = token.accessToken
      session.user.is_verified = token.is_verified
      session.user.profile_deep_link = token.profile_deep_link
      session.user.profile_web_link = token.profile_web_link
      return session
    },
  },
})

declare module "next-auth" {
  interface Session {
    accessToken?: string,
  }

  interface User {
    is_verified?: boolean,
    profile_deep_link?: string
    profile_web_link?: string

  }
}

declare module "next-auth/jwt" {
  interface JWT {
    accessToken?: string
    profile_deep_link?: string
    profile_web_link?: string
    is_verified?: boolean
  }
}
```

**Pull request merged**